doc/test-framework/writing-e2e-tests.md

```diff
@@ -0,0 +1,209 @@
+
+### TestCtx
+TestCtx is a local context that stores important information for each test, such
+as the namespace for that test and the finalizer (cleanup) functions. By handling
+namespace and resource initialization through TestCtx, we can make sure that all
+resources are properly handled and removed after the test finishes.
+
+## Walkthrough: Writing Tests
+In this section, we will be walking through writing the e2e tests of the sample
+[memcached-operator][memcached-sample].
+
+### Main Test
+The first step to writing a test is to create the `main_test.go` file. The `main_test.go`
+file simply calls the test framework's main entry that sets up the framework and then
+starts the tests. It should be pretty much identical for all operators. This is what it
+looks like for the memcached-operator:
+```
+package e2e
+
+import (
+    "testing"
+
+    f "github.com/operator-framework/operator-sdk/pkg/test"
+)
+
+func TestMain(m *testing.M) {
+    f.MainEntry(m)
+}
+```
+
+### Individual Tests
+#### 1. Import the framework
+Once MainEntry sets up the framework, it runs the remainder of the tests. First, make
+sure to import `testing`, the operator-sdk test framework (`pkg/test`) as well as your operator's libraries:
+```
+import (
+    "testing"
+
+    cachev1alpha1 "github.com/operator-framework/operator-sdk-samples/memcached-operator/pkg/apis/cache/v1alpha1"
+
+    framework "github.com/operator-framework/operator-sdk/pkg/test"
+)
+```
+
+#### 2. Register types with framework scheme
+The next step is to register your operator's scheme with the framework's dynamic client.
+To do this, you need to create a list struct for your custom resource and pass the list
+object and the object's addToScheme function to the framework's AddToFrameworkScheme
+function. For our example memcached-operator, it looks like this:
+```
+memcachedList := &cachev1alpha1.MemcachedList{
+    TypeMeta: metav1.TypeMeta{
+        Kind:       "Memcached",
+        APIVersion: "cache.example.com/v1alpha1",
+    },
+}
+err := framework.AddToFrameworkScheme(cachev1alpha1.AddToScheme, memcachedList)
+if err != nil {
+    t.Fatalf("failed to add custom resource scheme to framework: %v", err)
+}
+```
+
+The reason that the list struct is necessary for this function is because the dynamic client's RESTMapper
+needs to have the custom resource's mappings to work. These mappings are created by the kubernetes
+cluster. Since the CRD is created as the last step of the framework setup process, it is likely that the
+resource will not be ready by the time the first test starts. By providing the custom resource's list
+struct, the AddToFrameworkScheme function can verify that the dynamic client has the REST mappings for
+your custom resource. If the mappings are not yet available, the framework simply waits and attempts again
+1 second later. If the mappings are still not available after 5 seconds, the framework assumes something
+went wrong and returns an error.
```

Comment:
We can try to make this more concise. How about:

We pass in the CR List object `memcachedList` as an argument to `AddToFrameworkScheme()` because the framework needs to ensure that the dynamic client has the REST mappings to query the API server for the CR type. The framework will keep polling the API server for the mappings and timeout after 5 seconds to return an error if the mappings were not discovered in that time.